cleverhans/utils_tfe.py

```diff
@@ -0,0 +1,232 @@
+        nb_batches = int(math.ceil(float(len(X_train)) / args.batch_size))
+        assert nb_batches * args.batch_size >= len(X_train)
+
+        # Indices to shuffle training set
+        index_shuf = list(range(len(X_train)))
+        rng.shuffle(index_shuf)
+
+        prev = time.time()
+        for batch in range(nb_batches):
+
+            # Compute batch start and end indices
+            start, end = batch_indices(
+                batch, len(X_train), args.batch_size)
+
+            # Perform one training step
+            tf.assign(batch_x, X_train[index_shuf[start:end]])
+            tf.assign(batch_y, Y_train[index_shuf[start:end]])
+            # Compute grads
+            with tf.GradientTape() as tape:
+                # Define loss
+                loss_clean_obj = LossCrossEntropy(model, smoothing=0.)
+                loss_clean = loss_clean_obj.fprop(x=batch_x, y=batch_y)
+                loss = loss_clean
+                # Adversarial training
+                if attack is not None:
+                    batch_adv_x = attack.generate(batch_x, **attack_args)
+                    loss_adv_obj = LossCrossEntropy(model, smoothing=0.)
+                    loss_adv = loss_adv_obj.fprop(x=batch_adv_x, y=batch_y)
+                    loss = (loss_clean + loss_adv) / 2.0
+            # Apply grads
+            model_variables = model.get_params()
+            grads = tape.gradient(loss, model_variables)
+            optimizer.apply_gradients(zip(grads, model_variables))
+
+        assert end >= len(X_train)  # Check that all examples were used
+        cur = time.time()
+        _logger.info("Epoch " + str(epoch) + " took " +
+                     str(cur - prev) + " seconds")
+        if evaluate is not None:
+            evaluate()
+
+    if save:
+        save_path = os.path.join(args.train_dir, args.filename)
+        saver = tf.train.Saver()
+        saver.save(save_path, model_variables)
+        _logger.info("Completed model training and saved at: " +
+                     str(save_path))
+    else:
+        _logger.info("Completed model training.")
+
+    return True
+
+
+def model_eval(model, X_test=None, Y_test=None, args=None,
+               attack=None, attack_args=None):
+    """
+    Compute the accuracy of a TF Eager model on some data
+    :param model: instance of cleverhans.model.Model_Eager
+                    with pretrained weights for evaluation.
+    :param X_test: numpy array with training inputs
+    :param Y_test: numpy array with training outputs
+    :param args: dict or argparse `Namespace` object.
+                 Should contain `batch_size`
+    :param attack: instance of the class cleverhans.attacks.attacks_eager
+    :param attack_args: parameters required for the attack.
+    :return: a float with the accuracy value
+    """
+    args = _ArgsWrapper(args or {})
+
+    if ((attack is None) != (attack_args is None)):
+        error = "attack and attack_args must be passed together."
```

Comment:
Same comment about intermediate variable, it looks simpler to just pass the string to ValueError.